File: src/main/java/com/stainberg/keditview/KoalaImageView.java
package com.stainberg.keditview;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Matrix;
import android.support.media.ExifInterface;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewTreeObserver;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.RelativeLayout;

import com.facebook.drawee.view.SimpleDraweeView;

import java.util.List;

import static com.stainberg.keditview.UtilsKt.eventInView;

/**
 * Created by Stainberg on 7/5/17.
 */

public class KoalaImageView extends FrameLayout implements KoalaBaseCellView {
    private SimpleDraweeView imageView;
    private ImageView delete;
    private ImageView drag;
    private OnImageDeleteListener listener;
    private String src;
    private boolean visible;
    private int width, height;
    private Bitmap bitmap;
    private int bound = getResources().getDisplayMetrics().heightPixels;
    private int margin;

    public KoalaImageView(Context context) {
        this(context, null, 0);
    }

    public KoalaImageView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public KoalaImageView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private FileData fileData;

    @Deprecated
    public KoalaImageView(Context context, OnImageDeleteListener l, int w, int h) {
        super(context);
        width = w;
        height = h;
        listener = l;
        init();
    }

    public FileData getFileData() {
        return fileData;
    }

    public KoalaImageView(Context context, FileData fileData, OnImageDeleteListener l, int margin) {
        super(context);
        this.fileData = fileData;
        this.margin = margin;
        listener = l;
        src = TextUtils.isEmpty(fileData.fileUrl) ? fileData.filePath : fileData.fileUrl;
        init();
        reloadImage();
    }

    private boolean isDragEnabled = false;

    @Override
    public boolean onInterceptTouchEvent(MotionEvent ev) {
        if (null != drag && drag.getVisibility() == View.VISIBLE) {
            isDragEnabled = eventInView(ev, drag);
            if (isDragEnabled) {
                return isDragEnabled;
            }
        }
        return super.onInterceptTouchEvent(ev);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        if (isDragEnabled) {
            return false;
        }
        return super.onTouchEvent(event);
    }

    private void init() {
        final View v = LayoutInflater.from(getContext()).inflate(R.layout.item_view_image, this, true);
        delete = v.findViewById(R.id.icon_delete);
        delete.setOnClickListener(onDeleteImageListener);
        drag = v.findViewById(R.id.icon_drag);
        visible = false;
        setOnClickListener(onClickListener);
        getViewTreeObserver().addOnScrollChangedListener(onScrollChangedListener);
        width = fileData.width;
        height = fileData.height;
        float x, y;
        x = getResources().getDisplayMetrics().widthPixels - 2 * margin;
        y = x / ((float) width / (float) height);
        imageView = v.findViewById(R.id.icon);
        RelativeLayout.LayoutParams lp = new RelativeLayout.LayoutParams((int) x, (int) y);
        imageView.setLayoutParams(lp);
    }

    public String getFilePath() {
        return src;
    }

    @Override
    public String getUrl() {
        return src;
    }

    @Override
    public int getType() {
        return IMAGE_VIEW;
    }

    @Override
    public void reload() {
        int[] location = new int[2];
        imageView.getLocationInWindow(location);
        if (location[1] < 0) {
            if (location[1] < -(height + bound) && visible) {
                releaseImage();
                return;
            }
            if (location[1] > -(height + bound) && !visible) {
                reloadImage();
            }
        } else {
            if ((location[1] > getResources().getDisplayMetrics().heightPixels + bound && visible)) {
                releaseImage();
                return;
            }
            if ((location[1] < getResources().getDisplayMetrics().heightPixels + bound && !visible)) {
                reloadImage();
            }
        }
    }

    @Deprecated
    @Override
    public void setStyleH1() {

    }

    @Deprecated
    @Override
    public void setStyleH2() {

    }

    @Deprecated
    @Override
    public void setStyleNormal() {

    }

    @Deprecated
    @Override
    public void setGravity() {

    }

    @Deprecated
    @Override
    public void setQuote() {

    }

    @Deprecated
    @Override
    public void setSection(int st) {

    }

    @Deprecated
    @Override
    public void setBold() {

    }

    @Deprecated
    @Override
    public void setItalic() {

    }

    @Deprecated
    @Override
    public void setStrike() {

    }

    @Deprecated
    @Override
    public void addCode() {

    }

    @Deprecated
    @Override
    public void setText(CharSequence sequence) {

    }

    @Deprecated
    @Override
    public CharSequence getText() {
        return "";
    }

    @Deprecated
    @Override
    public List<String> getHtmlText() {
        return null;
    }

    @Override
    public void setHtmlText(String html) {

    }

    @Deprecated
    @Override
    public boolean isQuote() {
        return false;
    }

    @Deprecated
    @Override
    public boolean isCode() {
        return false;
    }

    @Override
    public int getStyle() {
        return 0;
    }

    @Override
    public int getSection() {
        return 0;
    }

    @Override
    public void setEditable(boolean enable) {

    }

    @Override
    public void enableDrag(boolean enable) {
        if (enable) {
            drag.setVisibility(VISIBLE);
        } else {
            drag.setVisibility(GONE);
        }
    }

    @Override
    public void release() {
        getViewTreeObserver().removeOnScrollChangedListener(onScrollChangedListener);
        releaseImage();
    }

    private void reloadImage() {
        System.out.println("reload bitmap");
        float w = width;
        float h = height;
        float rate = w / h;
        if (imageView.getAspectRatio() != rate) {
            imageView.setAspectRatio(w / h);
        }
        if (src.startsWith("http")) {
            imageView.setImageURI(src);
        } else {
            if (bitmap != null) {
                return;
            }
            KoalaImageLoadPoll.getPoll().handle(new Runnable() {
                @Override
                public void run() {
                    try {
                        BitmapFactory.Options options = new BitmapFactory.Options();
                        options.inJustDecodeBounds = true;
                        BitmapFactory.decodeFile(src, options);
                        int w = options.outWidth;
                        float scale = w / 800;
                        int s = (int) Math.ceil(scale);
                        options.inJustDecodeBounds = false;
                        options.inSampleSize = s;
                        bitmap = BitmapFactory.decodeFile(src, options);
                        int digree;
                        ExifInterface exif = new ExifInterface(src);
                        int ori = exif.getAttributeInt(ExifInterface.TAG_ORIENTATION,
                                ExifInterface.ORIENTATION_UNDEFINED);
                        switch (ori) {
                            case ExifInterface.ORIENTATION_ROTATE_90:
                                digree = 90;
                                break;
                            case ExifInterface.ORIENTATION_ROTATE_180:
                                digree = 180;
                                break;
                            case ExifInterface.ORIENTATION_ROTATE_270:
                                digree = 270;
                                break;
                            default:
                                digree = 0;
                                break;
                        }
                        if (digree != 0) {
                            Matrix m = new Matrix();
                            m.postRotate(digree);
                            bitmap = Bitmap.createBitmap(bitmap, 0, 0, bitmap.getWidth(), bitmap.getHeight(), m, true);
                            imageView.setImageBitmap(bitmap);
                        }
                        post(new Runnable() {
                            @Override
                            public void run() {
                                imageView.setImageBitmap(bitmap);
                            }
                        });
                    } catch (Exception e) {

                    }

                }
            });
        }
        visible = true;
    }

    public void releaseImage() {
        visible = false;
        imageView.setImageBitmap(null);
        imageView.setBackgroundColor(ContextCompat.getColor(getContext(), R.color.gray_placeholder));
        if (bitmap != null) {
            bitmap.recycle();
            bitmap = null;
        }
    }

    private OnClickListener onClickListener = new OnClickListener() {
        @Override
        public void onClick(View v) {
            if (delete.getVisibility() == VISIBLE) {
                delete.setVisibility(GONE);
            } else {
                delete.setVisibility(VISIBLE);
            }
        }
    };

    private OnClickListener onDeleteImageListener = new OnClickListener() {
        @Override
        public void onClick(View v) {
            listener.delete(KoalaImageView.this);
        }
    };

    ViewTreeObserver.OnScrollChangedListener onScrollChangedListener = new ViewTreeObserver.OnScrollChangedListener() {
        @Override
        public void onScrollChanged() {
            int[] location = new int[2];
            imageView.getLocationInWindow(location);
            if (location[1] < 0) {
                if (location[1] < -(height + bound) && visible) {
                    releaseImage();
                    return;
                }
                if (location[1] > -(height + bound) && !visible) {
                    reloadImage();
                }
            } else {
                if ((location[1] > getResources().getDisplayMetrics().heightPixels + bound && visible)) {
                    releaseImage();
                    return;
                }
                if ((location[1] < getResources().getDisplayMetrics().heightPixels + bound && !visible)) {
                    reloadImage();
                }
            }
        }
    };

    interface OnImageDeleteListener {
        void delete(KoalaBaseCellView v);
    }

}

File: src/main/java/com/stainberg/keditview/KoalaImageView.kt
package com.stainberg.keditview

import android.animation.ObjectAnimator
import android.content.Context
import android.net.Uri
import android.support.v4.content.ContextCompat
import android.text.TextUtils
import android.util.AttributeSet
import android.util.Log
import android.view.*
import android.widget.FrameLayout
import com.facebook.drawee.backends.pipeline.Fresco
import com.facebook.imagepipeline.common.ResizeOptions
import com.facebook.imagepipeline.request.ImageRequest
import com.facebook.imagepipeline.request.ImageRequestBuilder
import kotlinx.android.synthetic.main.item_view_edit_text.view.edit_text
import kotlinx.android.synthetic.main.item_view_file.view.*
import kotlinx.android.synthetic.main.item_view_image.view.*
import java.io.File
import java.lang.ref.SoftReference


/**
 * Created by Stainberg on 7/5/17.
 */

class KoalaImageView : FrameLayout, KoalaBaseCellView {
    var filePath: String = ""
    private var imgWidth: Int = 0
    private var imgHeight: Int = 0
    private val request: ImageRequest
    var fileData: FileData
    private var dragTouchToggled = false
    private var textStatus = 0
    private var isDragging = false
    private var isImageReleased = true

    private var sr: SoftReference<KoalaRichEditorView.Companion.IOnImageClickListener?>

    constructor(context: Context, fileData: FileData, lis: KoalaRichEditorView.Companion.IOnImageClickListener?) : super(context) {
        this.sr = SoftReference(lis)
        this.fileData = fileData
        filePath = if (TextUtils.isEmpty(fileData.fileUrl)) Uri.fromFile(File(fileData.filePath)).toString() else fileData.fileUrl
        init()

        var w: Float
        var h: Float
        if (fileData.width < fileData.height) {
            w = fileData.width.toFloat()
            while (w > 1000) {
                w -= 200
            }
            h = fileData.height / (fileData.width / w)
        } else {
            h = fileData.height.toFloat()
            while (h > 1000) {
                h -= 200
            }
            w = fileData.width.toFloat() / fileData.height.toFloat() * h
        }

        request = ImageRequestBuilder.newBuilderWithSource(Uri.parse(filePath))
                .setResizeOptions(ResizeOptions(w.toInt(), h.toInt()))
                .build()
        reloadImage()
    }

    override fun onInterceptTouchEvent(ev: MotionEvent): Boolean {
        if (image_icon_drag.visibility == View.VISIBLE) {
            dragTouchToggled = eventInView(ev, image_icon_drag)
            if (dragTouchToggled) {
                return dragTouchToggled
            }
        }
        return super.onInterceptTouchEvent(ev)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        return if (dragTouchToggled) {
            false
        } else super.onTouchEvent(event)
    }

    private fun init() {
        LayoutInflater.from(context).inflate(R.layout.item_view_image, this, true)
        image_left_area.setOnClickListener {
            if (!isDragging) {
                image_container.requestFocus()
                textStatus = 1
                updateTextStatus()
            }
        }
        image_right_area.setOnClickListener {
            if (!isDragging) {
                image_container.requestFocus()
                textStatus = 2
                updateTextStatus()
            }
        }
        image_center_area.setOnClickListener {
            textStatus = 0
            updateTextStatus()
            sr.get()?.onImageClick((parent as ViewGroup).indexOfChild(this))
        }
        image_container.setOnFocusChangeListener { _, hasFocus ->
            if (!hasFocus || isDragging) {
                textStatus = 0
                updateTextStatus()
            }
        }
        imgWidth = fileData.width
        imgHeight = fileData.height
        icon.measure(MeasureSpec.AT_MOST, MeasureSpec.UNSPECIFIED)
        val whRate = imgWidth.toFloat() / imgHeight.toFloat()
        val x: Float = context.screenWidth - context.dp2px(20f) * 2
        val y: Float = x / whRate

        var lp = image_content_bg.layoutParams
        lp.height = y.toInt()
        image_content_bg.layoutParams = lp

        image_container.setOnKeyListener { v, keyCode, event ->
            if (keyCode == KeyEvent.KEYCODE_BACK || KeyEvent.isModifierKey(keyCode)) {
                false
            } else {
                if (textStatus != 0 && !isDragging) {
                    if (event.action == KeyEvent.ACTION_UP || event.action == KeyEvent.ACTION_MULTIPLE) {
                        val index = (parent as ViewGroup).indexOfChild(this@KoalaImageView)
                        if (keyCode == KeyEvent.KEYCODE_DEL) {
                            if (textStatus == 1) {
                                val pre = KoalaRichEditorView.getPrev(parent as ViewGroup, this@KoalaImageView)
                                if (null != pre) {
                                    if (pre is KoalaEditTextView) {
                                        pre.edit_text.requestFocus()
                                        pre.edit_text.setSelection(pre.edit_text.length())
                                    } else if (pre is KoalaImageView) {
                                        pre.image_right_area.performClick()
                                    } else if (pre is KoalaFileView) {
                                        pre.file_right_area.performClick()
                                    }
                                }
                            } else if (textStatus == 2) {
                                deleteCurrentItem(index)
                            }
                        } else if (keyCode == KeyEvent.KEYCODE_ENTER) {
                            if (textStatus == 1) {
                                insertNewLine("", index)
                            } else if (textStatus == 2) {
                                insertNewLine("", index + 1)
                            }
                        } else {
                            val s: String? = if (keyCode != 0) {
                                if (event != null && event.unicodeChar != null) {
                                    event.unicodeChar.toChar().toString()
                                } else {
                                    null
                                }
                            } else {
                                if (null != event && null != event.characters) {
                                    event.characters.toString()
                                } else {
                                    null
                                }
                            }
                            s?.let {
                                if (textStatus == 1) {
                                    insertNewLine(s, index)
                                } else if (textStatus == 2) {
                                    insertNewLine(s, index + 1)
                                }
                            }
                        }
                    }
                }
                true
            }
        }
    }

    private fun updateTextStatus() {
        val leftVisibility = if (textStatus == 1) VISIBLE else GONE
        if (image_left_line.visibility != leftVisibility) {
            image_left_line.visibility = leftVisibility
        }
        val rightVisibility = if (textStatus == 2) VISIBLE else GONE
        if (image_right_line.visibility != rightVisibility) {
            image_right_line.visibility = rightVisibility
        }
        if (textStatus != 0) {
            post { showSoft() }
        }
    }

    override fun resetMargin() {
        initMargin()
    }

    private val offset = context.screenHeight
    private val globalLayoutListener = ViewTreeObserver.OnGlobalLayoutListener {
        val height = image_container.height
        if (height != 0) {
            val lp = image_space.layoutParams
            if (lp.height != height) {
                lp.height = height
                image_space.layoutParams = lp
            }
        }
    }
    private val scrollChangedListener = ViewTreeObserver.OnScrollChangedListener {
        val location = IntArray(2)
        getLocationInWindow(location)
        val y = location[1]
        if (y > offset * 2 || (y + height < -offset)) {
            if (image_container.visibility != View.GONE) {
                image_container.visibility = View.GONE
            }
            releaseImage()
        } else {
            if (image_container.visibility != View.VISIBLE) {
                image_container.visibility = View.VISIBLE
            }
            reloadImage()
        }
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        viewTreeObserver.addOnGlobalLayoutListener(globalLayoutListener)
        viewTreeObserver.addOnScrollChangedListener(scrollChangedListener)
        reloadImage()
        initMargin()
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        viewTreeObserver.removeOnGlobalLayoutListener(globalLayoutListener)
        viewTreeObserver.removeOnScrollChangedListener(scrollChangedListener)
        releaseImage()
    }

    private fun initMargin() {
        val pre = KoalaRichEditorView.getPrev(parent as ViewGroup, this)
        pre?.let {
            val lp = image_content_bg.layoutParams as MarginLayoutParams
            when (pre) {
                is KoalaImageView, is KoalaFileView -> {
                    lp.topMargin = MARGIN_4
                }
                is KoalaEditTextView -> {
                    lp.topMargin = MARGIN_3
                }
                else -> {
                    lp.topMargin = MARGIN_5
                }
            }
            image_content_bg.layoutParams = lp
        }
    }

    private val defaultPadding = context.dp2px(4f).toInt()
    private val paddingAnim: PaddingAnim by lazy { PaddingAnim(image_content_bg) }
    override fun enableDrag(enable: Boolean) {
        isDragging = enable
        textStatus = 0
        updateTextStatus()
        image_container.showShadow(enable)
        if (enable) {
            image_container.isFocusable = false
            image_container.isFocusableInTouchMode = false
            ObjectAnimator.ofInt(paddingAnim, "padding", image_content_bg.paddingLeft, defaultPadding).setDuration(animTime).start()
            image_icon_drag.visibility = View.VISIBLE
            image_touch_container.visibility = View.GONE
        } else {
            image_container.isFocusable = true
            image_container.isFocusableInTouchMode = true
            ObjectAnimator.ofInt(paddingAnim, "padding", image_content_bg.paddingLeft, 0).setDuration(animTime).start()
            image_icon_drag.visibility = View.GONE
            image_touch_container.visibility = View.VISIBLE
        }
    }

    fun actionDown() {
        image_icon_drag.setImageResource(R.drawable.svg_drag_icon_selected)
        image_container.showHighLight(true)
    }

    fun actionUp() {
        image_icon_drag.setImageResource(R.drawable.svg_drag_icon)
        image_container.showHighLight(false)
    }

    private fun reloadImage() {
        if (!isImageReleased) return
        isImageReleased = !isImageReleased
        icon.controller = Fresco.newDraweeControllerBuilder().setImageRequest(request).build()
    }

    private fun releaseImage() {
        if (isImageReleased) return
        isImageReleased = !isImageReleased
        icon.setImageBitmap(null)
        val imagePipeline = Fresco.getImagePipeline()
        val uri: Uri = Uri.fromFile(File(filePath))
        imagePipeline.evictFromMemoryCache(uri)
    }

    private fun insertNewLine(text: String, index: Int) {
        val krev = parent?.parent?.parent?.parent ?: return
        if (krev is KoalaRichEditorView) {
            krev.addCellText(text, index)
        }
    }

    private fun deleteCurrentItem(index: Int) {
        val krev = parent?.parent?.parent?.parent ?: return
        if (krev is KoalaRichEditorView) {
            krev.deleteImage(index)
            krev.addCellText("", index)
        }
    }
}
